portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -14,24 +14,52 @@
 
 package com.google.sps.servlets;
 
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.gson.Gson;
+import com.google.sps.data.Comment;
 import java.io.IOException;
+import java.lang.Integer;
+import java.text.SimpleDateFormat;
+import java.util.ArrayList;
+import java.util.Date;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
-import com.google.gson.Gson;
-import java.util.ArrayList;
-
 /** Servlet that returns some example content. TODO: modify this file to handle comments data */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
-
-  public ArrayList<String> comments = new ArrayList<String>();
   
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
 
+    Query query = new Query(Comment.MESSAGE_KEY).addSort(Comment.TIMESTAMP_KEY, SortDirection.DESCENDING);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    int count = 0;
+    ArrayList<Comment> comments = new ArrayList<Comment>();
+
+    for(Entity entity : results.asIterable()) {
+      String name = (String) entity.getProperty(Comment.NAME_KEY);
+      String message = (String) entity.getProperty(Comment.MESSAGE_KEY);
+      long timestamp = (long) entity.getProperty(Comment.TIMESTAMP_KEY);
+      String time = (String) entity.getProperty(Comment.TIME_KEY);
+
+      Comment comment = new Comment(name, message, timestamp, time);
+      comments.add(comment); 
+      count ++;
+      if (count == 100);
```

Comment:
I guess the instruction has not given this information (But I want to let you know since it is helpful). You don't have to do it. You could potentially 
- fetch a fixed/max amount of items from the database. [See](https://cloud.google.com/appengine/docs/standard/java/javadoc/com/google/appengine/api/datastore/PreparedQuery.html#countentities). This will increase the efficiency a lot without keeping a counter inside the iteration. For example, if the database has 1K items and you just need 10 of them. A full lookup of 1000 items will cost 100 times more than a look up with 10 items back, and you don't have to count yourself afterwords.
Here is an example:
```
import com.google.appengine.api.datastore.FetchOptions;

 Cursor cursor = ...

 ...

 // limit 10
 datastoreService.prepare(query).asList(FetchOptions.Builder.withLimit(10));

 // limit 10, start cursor
 datastoreService.prepare(query).asList(FetchOptions.Builder.withLimit(10).startCursor(cursor));
 ```
- get the result as a list but not only iterable, and there are other options as well. [See](https://cloud.google.com/appengine/docs/standard/java/javadoc/com/google/appengine/api/datastore/PreparedQuery#aslist).